lib/spack/spack/build_systems/intel.py

```diff
@@ -61,132 +110,1028 @@ class IntelPackage(PackageBase):
+    @property
+    def headers(self):
+        if '+mpi' in self.spec or self.provides('mpi'):
+            return find_headers(
+                ['mpi'],
+                root=self.component_include_dir('mpi'),
+                recursive=False)
+        if '+mkl' in self.spec or self.provides('mkl'):
+            return find_headers(
+                ['mkl_cblas', 'mkl_lapacke'],
+                root=self.component_include_dir('mkl'),
+                recursive=False)
+
+    @property
+    def libs(self):
+        if '+mpi' in self.spec or self.provides('mpi'):
+            # If prefix is too general, recursive searches may get files from
+            # supported but inappropriate sub-architectures like 'mic'.
+            libnames = ['libmpifort', 'libmpi']
+            if 'cxx' in self.spec.last_query.extra_parameters:
+                libnames = ['libmpicxx'] + libnames
+            return find_libraries(
+                libnames,
+                root=self.component_lib_dir('mpi'),
+                shared=True, recursive=True)
+
+    def setup_environment(self, spack_env, run_env):
+        """Adds environment variables to the generated module file.
+
+        These environment variables come from running:
+
+        .. code-block:: console
+
+           $ source parallel_studio_xe_2017/bin/psxevars.sh intel64
+           [and likewise for MKL, MPI, and other components]
+        """
+        # NOTE: Spack runs setup_environment twice, once pre-build to set up
+        # the build environment, and once post-installation to determine
+        # the environment variables needed at run-time to add to the module
+        # file. The script we need to source is only present post-installation,
+        # so check for its existence before sourcing.
+        # TODO: At some point we should split setup_environment into
+        # setup_build_environment and setup_run_environment to get around
+        # this problem.
+        f = self.file_to_source
+        if not f or not os.path.isfile(f):
+            return
+
+        tty.debug("sourcing " + f)
+
+        if sys.platform == 'darwin':
+            args = ()
+        else:
+            # All Intel packages expect at least the architecture as argument.
+            # Some accept more args, but those are not (yet?) handled here.
+            args = (_expand_fields('{arch}'),)
+
+        run_env.extend(EnvironmentModifications.from_sourcing_file(f, *args))
+
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        if '+mkl' in self.spec or self.provides('mkl'):
+            # Spack's env philosophy demands that we replicate some of the
+            # settings normally handled by file_to_source ...
+            spack_env.set('MKLROOT', self.normalize_path('mkl'))
+            spack_env.append_path('SPACK_COMPILER_EXTRA_RPATHS',
+                                  self.component_lib_dir('mkl'))
+        # For mpi, see mpi_setup_dependent_environment()
```

Comment:
This function is overridden in `IntelParallelStudio`. Does this one even get called then?

Reply:
Right, it didn't. – I now funnel both the class version of `setup_dependent_environment()` and the client-specific ones through a `_setup_dependent_env_callback()` function instead.